Implements all methods of the ENS registry (#3325)
ethereum/web3.js: docs/web3-eth-ens.rst

.. _eth-ens:

> ============

Author: Changed because it is otherwise invalid.

------------------------------------------------------------------------------

getResolver
===========

.. code-block:: javascript

    web3.eth.ens.getResolver(name [, callback]);

Returns the resolver contract to an Ethereum address.

----------
Parameters
----------

1. ``name`` - ``String``: The ENS name.
2. ``callback`` - ``Function``: (optional) Optional callback

-------
Returns
-------

``Promise<Resolver>`` - The ENS resolver for this name.

-------
Example
-------

.. code-block:: javascript

    web3.eth.ens.getResolver('ethereum.eth').then(function (contract) {
        console.log(contract);
    });
    > Contract<Resolver>

------------------------------------------------------------------------------

setResolver
===========

.. code-block:: javascript

    web3.eth.ens.setResolver(name, address [, txConfig ] [, callback]);

> Does set the resolver contract address of a name.

Reviewer: ```suggestion
Sets the resolver contract address of a name
```

4. ``callback`` - ``Function``: (optional) Optional callback

-------
Returns
-------

``PromiEvent<TransactionReceipt | TransactionRevertInstructionError>``

-------
Example
-------

.. code-block:: javascript

    web3.eth.ens.setResolver('ethereum.eth', '0x...', {...}).then(function (receipt) {
        console.log(receipt);
    });
    > {...}

------------------------------------------------------------------------------

getOwner
========

.. code-block:: javascript

    web3.eth.ens.getOwner(name [, callback]);

Returns the owner of a name.

----------
Parameters
----------

1. ``name`` - ``String``: The ENS name.
2. ``callback`` - ``Function``: (optional) Optional callback

-------
Returns
-------

`Promise<String>`` - The address of the registrar (EOA or CA).

-------
Example
-------

.. code-block:: javascript

    web3.eth.ens.getOwner('ethereum.eth').then(function (owner) {
        console.log(owner);
    });
    > '0x...'

------------------------------------------------------------------------------

setOwner
========

.. code-block:: javascript

    web3.eth.ens.setOwner(name [, txConfig ] [, callback]);

> Does set the owner of the given name.

Reviewer: ```suggestion
Sets the owner of the given name.
```

3. ``callback`` - ``Function``: (optional) Optional callback

-------
Returns
-------

``PromiEvent<TransactionReceipt | TransactionRevertInstructionError>``

-------
Example
-------

.. code-block:: javascript

    web3.eth.ens.setOwner('ethereum.eth', {...}).then(function (receipt) {
        console.log(receipt);
    });
    > {...}

------------------------------------------------------------------------------

getTTL
======

.. code-block:: javascript

    web3.eth.ens.getTTL(name [, callback]);

Returns the caching TTL (time-to-live) of a name.

----------
Parameters
----------

1. ``name`` - ``String``: The ENS name.
2. ``callback`` - ``Function``: (optional) Optional callback

-------
Returns
-------

``Promise<Number>``

-------
Example
-------

.. code-block:: javascript

    web3.eth.ens.getTTL('ethereum.eth').then(function (ttl) {
        console.log(ttl);
    });
    > 100000

------------------------------------------------------------------------------

setTTL
======

.. code-block:: javascript

    web3.eth.ens.setTTL(name, ttl [, txConfig ] [, callback]);

> Does set the caching TTL (time-to-live) of a name.

Reviewer: ```suggestion
Sets the caching TTL (time-to-live) of a name.
```

-------
Example
-------

.. code-block:: javascript

    web3.eth.ens.setTTL('ethereum.eth', 10000, {...}).then(function (receipt) {
        console.log(receipt);
    });
    > {...}

------------------------------------------------------------------------------

setSubnodeOwner
===============

.. code-block:: javascript

    web3.eth.ens.setSubnodeOwner(name, label, address [, txConfig ] [, callback]);

Creates a new subdomain of the given node, assigning ownership of it to the specified owner

----------
Parameters
----------

1. ``name`` - ``String``: The ENS name.
2. ``label`` - ``String``: The name of the sub-domain or the sha3 hash of it
3. ``address`` - ``String``: The registrar of this sub-domain
4. ``txConfig`` - ``Object``: (optional) The transaction options as described ::ref::`here <eth-sendtransaction>`
5. ``callback`` - ``Function``: (optional) Optional callback

-------
Returns
-------

``PromiEvent<TransactionReceipt | TransactionRevertInstructionError>``

-------
Example
-------

.. code-block:: javascript

    web3.eth.ens.setSubnodeOwner('ethereum.eth', 'web3', '0x...', {...}).then(function (receipt) {

>         console.log(receipt); // successfully web3.ethereum.eth registered

Reviewer: ```suggestion
        console.log(receipt); // successfully registered web3.ethereum.eth
```

    });
    > {...}

------------------------------------------------------------------------------

setRecord
=========

.. code-block:: javascript

    web3.eth.ens.setRecord(name, owner, resolver, ttl, [, txConfig ] [, callback]);

Sets the owner, resolver, and TTL for an ENS record in a single operation.

----------
Parameters
----------

1. ``name`` - ``String``: The ENS name.
2. ``owner`` - ``String``: The owner of the name record
3. ``resolver`` - ``String``: The resolver address of the name record
4. ``ttl`` - ``String | Number``: Time to live value (uint64)
5. ``txConfig`` - ``Object``: (optional) The transaction options as described ::ref::`here <eth-sendtransaction>`
6. ``callback`` - ``Function``: (optional) Optional callback

-------
Returns
-------

``PromiEvent<TransactionReceipt | TransactionRevertInstructionError>``

-------
Example
-------

.. code-block:: javascript

    web3.eth.ens.setRecord('ethereum.eth', '0x...', '0x...', 1000000, {...}).then(function (receipt) {

>         console.log(receipt); // successfully web3.ethereum.eth registered

Reviewer: ```suggestion
        console.log(receipt); // successfully registered web3.ethereum.eth 
```

Is this comment valid for all three places it appears?
+ setSubnodeOwner
+ setRecord
+ setSubnodeRecord

Author: Oh, thanks! I will remove it:)

Author: The inline comment in the example is correct for the [setSubnodeOwner](https://docs.ens.domains/contract-api-reference/ens#set-subdomain-owner) and [setSubnodeRecord](https://docs.ens.domains/contract-api-reference/ens#set-subdomain-record) method.

I've changed the inline comment of ``setSubnodeOwner`` to: 
``// successfully defined the owner of web3.ethereum.eth``

    });
    > {...}

------------------------------------------------------------------------------

setSubnodeRecord
================

.. code-block:: javascript

    web3.eth.ens.setSubnodeRecord(name, label, owner, resolver, ttl, [, txConfig ] [, callback]);

Sets the owner, resolver and TTL for a subdomain, creating it if necessary.

----------
Parameters
----------

1. ``name`` - ``String``: The ENS name.
2. ``label`` - ``String``: The name of the sub-domain or the sha3 hash of it
3. ``owner`` - ``String``: The owner of the name record
4. ``resolver`` - ``String``: The resolver address of the name record
5. ``ttl`` - ``String | Number``: Time to live value (uint64)
6. ``txConfig`` - ``Object``: (optional) The transaction options as described ::ref::`here <eth-sendtransaction>`
7. ``callback`` - ``Function``: (optional) Optional callback

-------
Returns
-------

``PromiEvent<TransactionReceipt | TransactionRevertInstructionError>``

-------
Example
-------

.. code-block:: javascript

    web3.eth.ens.setSubnodeRecord('ethereum.eth', 'web3', '0x...', '0x...', 1000000, {...}).then(function (receipt) {

>         console.log(receipt); // successfully web3.ethereum.eth registered

Reviewer: ```suggestion
        console.log(receipt); // successfully registered web3.ethereum.eth
```

-------
Example
-------

.. code-block:: javascript

    web3.eth.ens.setApprovalForAll('0x...', true, {...}).then(function (receipt) {
        console.log(receipt);
    });
    > {...}

------------------------------------------------------------------------------

isApprovedForAll
================

.. code-block:: javascript

    web3.eth.ens.isApprovedForAll(owner, operator [, callback]);

Returns ``true`` if the operator is approved to make ENS registry operations on behalf of the owner.

----------
Parameters
----------

1. ``owner`` - ``String``: The owner address.
2. ``operator`` - ``String``: The operator address.
3. ``callback`` - ``Function``: (optional) Optional callback

-------
Returns
-------

``Promise<Boolean>``

-------
Example
-------

.. code-block:: javascript

    web3.eth.ens.isApprovedForAll('0x0...', '0x0...').then(function (isApproved) {
        console.log(isApproved);
    })
    > true

------------------------------------------------------------------------------

recordExists
============

.. code-block:: javascript

    web3.eth.ens.recordExists(name [, callback]);

Returns ``true`` if node exists in this ENS registry.

> This will return ``false`` for records that are in the legacy ENS registry but have not yet been migrated to the new one.

Reviewer: 👍

.. code-block:: javascript

>     web3.eth.ens.setAddress(ENSName, address [, txConfig ] [, callback]);

Author: Optional parameters weren't documented correctly for already implemented methods.

1. ``ENSName`` - ``String``: The ENS name.
2. ``address`` - ``String``: The address to set.

> 3. ``txConfig`` - ``Object``: (optional) The transaction options as described ::ref::`here <eth-sendtransaction>`

Author: It's better to forward them to the ``sendTransaction`` documentation instead of duplicating the transaction options documentation here.
